src/platform_storage_api/admission_controller/api.py

```diff
@@ -0,0 +1,219 @@
+            )
+
+        annotations = metadata.get("annotations", {})
+        raw_injection_spec = annotations.get(ANNOTATION_APOLO_INJECT_STORAGE)
+
+        if not raw_injection_spec:
+            # a pod does not request storage. we can do early-exit here
+            logger.info("POD won't be mutated because doesnt define proper annotations")
+            return admission_review.to_response()
+
+        return await self._handle_injection(
+            pod=pod,
+            raw_injection_spec=raw_injection_spec,
+            admission_review=admission_review,
+        )
+
+    async def _handle_new_platform_storage_pod(
+        self,
+        pod: dict[str, Any],
+        admission_review: AdmissionReviewResponse,
+    ) -> web.Response:
+        await self._volume_resolver.refresh_internal_state(pod=pod)
+        return admission_review.to_response()
+
+    async def _handle_injection(
+        self,
+        pod: dict[str, Any],
+        raw_injection_spec: str,
+        admission_review: AdmissionReviewResponse,
+    ) -> web.Response:
+        pod_spec = pod["spec"]
+        containers = pod_spec.get("containers") or []
+
+        if not containers:
+            logger.info("POD won't be mutated because doesnt define containers")
+            # pod does not define any containers. we can exit
+            return admission_review.to_response()
+
+        logger.info("Going to inject volumes")
+        try:
+            injection_spec = json.loads(raw_injection_spec)
+        except ValueError:
+            logger.info("Invalid injection spec. Denying the request")
+            admission_review.allowed = False
+            return admission_review.to_response()
+
+        # let's ensure POD has volumes
+        if "volumes" not in pod_spec:
+            admission_review.add_patch(
+                path="/spec/volumes",
+                value=[]
+            )
+
+        # and ensure that each container has a volume mounts
+        for idx, container in enumerate(containers):
+            if "volumeMounts" not in container:
+                admission_review.add_patch(
+                    path=f"/spec/containers/{idx}/volumeMounts",
+                    value=[]
+                )
+
+        for mount_path, storage_path in injection_spec.items():
+            # now let's try to resolve a path which POD wants to mount
+            try:
+                volume_spec = await self._volume_resolver.resolve_to_mount_volume(
+                    path=storage_path
+                )
+            except VolumeResolverError:
+                # report an error and disallow spawning a POD
+                logger.exception("Unable to resolve a volume for a provided path")
+                admission_review.allowed = False
+                return admission_review.to_response()
+
+            future_volume_name = create_injection_volume_name()
+
+            # add a volume host path
+            admission_review.add_patch(
```

Comment:
There can be multiple volumes. For example user wants to mount storages of different orgs and copy data between them.